### ISSUE TITLE
ZDI-reported vulnerability

### ISSUE THREAD

**Issue by sean-geggie-keyshot** [alembic/alembic#453]
We have been advised by Zero Day Initiative that our usage of the Alembic framework may constitute a remote code execution vulnerability. They have sent us the attached file illustrating the issue. Indeed, we see an access violation when running the file through our importer. Passing through the samples included with the framework gives similar results.

[zdi-23700-poc0.zip](https://github.com/user-attachments/files/17130050/zdi-23700-poc0.zip)

**Comment by lamiller0** [comment]
Thank you very much for reporting this. 

I see this was written with Alembic 1.7.7, is this the version you are using for your Blender importer?

We have fixed several similiar reported vulnerabilities in the past, and added them to our test suite here:
https://github.com/alembic/alembic/tree/master/lib/Alembic/AbcCoreOgawa/Tests

Do you still get the access violation when using Alembic 1.8.5?

**Comment by sean-geggie-keyshot** [comment]
We are testing with 1.8.6. Testing with Address Sanitizer enabled shows a heap buffer overflow, which we assume is related. This can be reproduced simply by, for instance, adding something like the following method to PolyMeshTest.cpp:
```
void ZDITest()
{
  IArchive archive(Alembic::AbcCoreOgawa::ReadArchive(), "zdi-23700-poc1.abc");
  auto top = archive.getTop();
  // recursive call to children
  for (size_t i = 0; i < top.getNumChildren(); i++) {
    auto child = top.getChild(i);
    for (size_t j = 0; j < child.getNumChildren(); j++) {
      auto childChild = child.getChild(j);
      std::string schemaName = childChild.getMetaData().get("schema");
      if (schemaName == ALEMBIC_ABCGEOM_POLYMESH_SCHEMA) {
        IPolyMesh childChildMesh(child, childChild.getName());
        IPolyMeshSchema& mesh = childChildMesh.getSchema();
        IPolyMeshSchema::Sample mesh_samp;
        mesh.get(mesh_samp);
        Int32ArraySamplePtr meshFaceIndices = mesh_samp.getFaceIndices();
        IN3fGeomParam normals = mesh.getNormalsParam();
        auto valueSampler = normals.getValueProperty().getValue();
      }
    }
  }
}
```
This is a very boiled down version of our own code which converts alembic files into our own scene format.

With address sanitizer enabled, we see a heap-buffer-overflow in `mesh.get(mesh_samp)`, although it is not clear to me why.
We have an additional file which gives a heap-buffer-overflow in `getValue()` in the same example: [zdi-23700-poc1.zip](https://github.com/user-attachments/files/17160811/zdi-23700-poc1.zip)

**Comment by lamiller0** [comment]
Thanks for the hints.  I'm able to reproduce this and am looking into a fix.

**Comment by lamiller0** [comment]
I have a partial solution that I'm not yet totally happy with in #454 

There was a mismatch between the size of the buffer we were reading, and the advertised size of the dimensions.

This was causing the buffer overflow.  I need to investigate if I can detect this in ReadDimensions and potentially throw an exception in there to give a hint about a malformed file.

**Comment by lamiller0** [comment]
This should hopefully be fixed in Alembic 1.8.7

**Comment by seangeggie** [comment]
Thanks so much!
